src/mame/includes/pc9801.h

```diff
@@ -440,13 +472,402 @@ class pc9801_state : public driver_device
 		bool init;
 	} m_egc;
 
-	uint16_t m_pc9821_256vram_bank;
+protected:
+	void egc_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+private:
+	void egc_blit_w(uint32_t offset, uint16_t data, uint16_t mem_mask);
+	uint16_t egc_blit_r(uint32_t offset, uint16_t mem_mask);
 
-	uint8_t m_sdip_read(uint16_t port, uint8_t sdip_offset);
-	void m_sdip_write(uint16_t port, uint8_t sdip_offset,uint8_t data);
 	uint16_t egc_do_partial_op(int plane, uint16_t src, uint16_t pat, uint16_t dst) const;
 	uint16_t egc_shift(int plane, uint16_t val);
 	uint16_t egc_color_pat(int plane) const;
+
+protected:
+	u8 mouse_freq_r(offs_t offset);
+	void mouse_freq_w(offs_t offset, u8 data);
+	
+	u8 ppi_mouse_portb_r();
+	u8 ppi_mouse_portc_r();
+};
+
+class pc9801us_state : public pc9801vm_state
+{
+public:
+	pc9801us_state(const machine_config &mconfig, device_type type, const char *tag)
+		: pc9801vm_state(mconfig, type, tag)
+	{
+	}
+	void pc9801us(machine_config &config);
+
+protected:
+	void pc9801us_io(address_map &map);
+
+	DECLARE_MACHINE_START(pc9801us);
+
+
+	// SDIP, PC9801DA onward
+protected:
+	u8 m_sdip[24];
+private:
+	u8 m_sdip_bank;
+	template<unsigned port> u8 sdip_r(offs_t offset);
+	template<unsigned port> void sdip_w(offs_t offset, u8 data);
+	void sdip_bank_w(offs_t offset, u8 data);
+};
+
+/**********************************************************
+ *
+ * BX class ("98Fellow")
+ *
+ **********************************************************/
+
+class pc9801bx_state : public pc9801us_state
+{
+public:
+	pc9801bx_state(const machine_config &mconfig, device_type type, const char *tag)
+		: pc9801us_state(mconfig, type, tag)
+	{
+	}
+
+	void pc9801bx2(machine_config &config);
+
+protected:
+	void pc9801bx2_io(address_map &map);
+	void pc9801bx2_map(address_map &map);
+
+	DECLARE_MACHINE_START(pc9801bx2);
+	DECLARE_MACHINE_RESET(pc9801bx2);
+
+private:
+	u8 i486_cpu_mode_r(offs_t offset);
+	u8 gdc_31kHz_r(offs_t offset);
+	void gdc_31kHz_w(offs_t offset, u8 data);
+};
+
+/******************************************
+ *
+ * pc9821.cpp: PC-9821 base class
+ *
+ *****************************************/
+
+class pc9821_state : public pc9801bx_state
```

Comment:
The declarations for the variants that are no longer in pc9801.cpp should be moved out of this header.  Either put them in headers corresponding to the source files if they’re going to be used as base classes further downstream, or move them to the source files themselves if that’s not going to be necessary.

Reply:
I choose option 1: 2 sounds too much bloat for stuff that eventually needs to be moved around when emulation evolves even more.